### PR TITLE
Support serialized checkpoint loading [redo #9605]

### LINKED ISSUE #9406: Support serialized checkpoint loading
### Motivation

Currently, all processes load the checkpoint at the same time. This can quickly OOM for large models. These use cases, especially with things like mixture of experts, might require serialized loading of checkpoint dicts across ranks (ie load the checkpoint one rank at a time per node). Could we enable this for DDP? 

Prior work: https://github.com/PyTorchLightning/pytorch-lightning/pull/8515

### Pitch

This would be controlled per training type plugin. Example pseudocode: https://gist.github.com/ananthsub/4ceedff56b2049a63bbb05ccd283b919

To work through: 
1. Should the TrainingTypePlugin have responsibility of calling `on_load_checkpoint` instead of the Trainer/connector? This would make sense as the TTP "owns" the LightningModule inside of the trainer, and since it already offers `load_model_state_dict`: https://github.com/PyTorchLightning/pytorch-lightning/blob/41ba639859cf6c6bf319eb33e5b3394504315962/pytorch_lightning/plugins/training_type/training_type_plugin.py#L159-L160

DeepSpeed already eschews most of the checkpoint connector logic when it comes to loading the lightning module state. This could be a gap for metrics, and this means we could be calling `on_load_checkpoint` multiple times with certain plugins. In my opinion, this points to all LightningModule state load/save/alterations should sit inside of the training type plugin. 

2. Does loading one process per node run into other issues with collectives potentially being used in `LightningModule.on_load_checkpoint` ?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #9406: redo #9605 now that #10061 resolves #10057.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -537,3 +544,29 @@ def teardown(self) -> None:
...
+                checkpoint = super().load_checkpoint(checkpoint_path)
+                self.lightning_module.on_load_checkpoint(checkpoint)
+                # call hpc specific hook
+                if self.lightning_module.trainer.checkpoint_connector.hpc_resume_path is not None:
+                    self.lightning_module.on_hpc_load(checkpoint)
+                self.load_model_state_dict(checkpoint)
+                del checkpoint["state_dict"]
+                self._has_loaded_state_dict = True
+                log.info(f"Rank {self.global_rank}: done loading model states from {checkpoint_path}.")
+            self.barrier()
```
[the remark this answers is not in the document]
Comment (yifuwang) [reply pytorch_lightning/plugins/training_type/ddp.py:573]: @awaelchli - the high level logic we have in mind is as follows:
- Before loading checkpoint, perform a collective to check if all ranks are loading/loading from the same path
  - If not, proceed with the regular loading logic
  - If yes:
    - Are we training on CPU?
      - If yes, proceed with the regular loading logic (there's no way to optimize for RAM footprint for this case)
    - Create local process groups consisting of ranks within the same host
    - Local rank 0 loads checkpoint from storage
    - Local rank 0 enables shared memory on tensors in `state_dict`
    - Local rank 0 serializes `state_dict` with `ForkingPickler` (the pickler used by `torch.multiprocessing.Queue` which knows how to handle shared memory tensors)
    - Broadcast the serialized `state_dict` with the local process group
    - Local rank 1-n deserializes `state_dict`
    - All ranks restore model with `state_dict`
    - Reclaim RAM occupied by the shared memory tensors

With this approach, we reduce storage I/O and RAM footprint by `[NUM_DEVICES_PER_HOST]` for all `dpp` use cases on `GPU`. Let me know what you think.

cc: @ananthsub

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp.py:573]: That sounds great @yifuwang !!

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:560]: This could also resolve the current bug where the checkpoint doesn't exist on some nodes when the filesystem isn't shared.

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:560]: @yifuwan, Would you mind opening an issue for this proposal?

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:560]: Should we close this PR and use the proposed logic ?

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:560]: @yifuwang @jjenniferdai Who would make a PR with the above proposition?

**Comment by tchaton** [comment]
@jjenniferdai What is the status with this PR ?

**Comment by jjenniferdai** [comment]
if everyone's onboard I think we can close this PR, I'll work with Yifu on the proposed alternative
